### PR TITLE
feat: Add Workforce

### PR DESCRIPTION
**Pull request by yiyiyi0817**
## Description

Consider one layer structure first, big workforce containing small leaf workforces + agents.

## Motivation and Context

A workflow/workforce for task completing.

1. Workforce(task="Human Task",society=[Agent A, Agent B, RolePlaying …..],manager_agent_config={…},
task_agent_config={…},… workforce = [], description=”xx”)→ ChatResult & Message
2. Guohao: Workforce(task=””, workforce=[workforce, …])
3. Agents in Workforce
    1. Manager Agent(optional): track task and can create new agent
    2. Task Agent(optional): Specify, Spilt, Generate, Plan tasks, already supported in CAMEL
    3. Critic Agent(optional): Assists in selecting an option, already supported in CAMEL
4. Flow in Workforce
    1. Human Task → Task A, Task B, Task C (Message) by Task Agent
    2. Manager Agent initialize the first round task assignment
        
        Task A -> Agent A (model=Model A, tool=[T1, T2],knowledgebase=[KB1, KB2])
        Task B -> Agent B (model=Model B, tool=[T1, T3, T5],knowledgebase=[KB2, KB3, KB4])
        Task C .... 
        
    3. environment for Agent A = related tasks information + all agents information
    environment for Manager Agent = all tasks information + all agents information
    4. If Agent A can solve task A, CAMEL Task A done, back to Manager Agent.
        
        If Agent A can’t solve task A, Agent A will ask Task Agent split Task A, make it into Task A.1, Task A.2…, from Agent A’s environment information, Agent A find another agent can solve Task A.1, send message to that Agent, assign Task A.1, once finished ,back to Manager Agent.
        
        Society pipeline? sequential, hierarchical, star, or other topology..  
        
        Task A.1 → Workforce with 3 agents

- [x] I have raised an issue to propose this change ([required](https://github.com/camel-ai/camel/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [x] Example (update in the folder of example)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/camel-ai/camel/blob/master/CONTRIBUTING.md) guide. (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly. (*required for a bug fix or a new feature*)
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: camel/workforce/workforce.py
Comment (willshang76) [inline camel/workforce/workforce.py:89]: I would see what is plan for the manager to decide what func to do. I think that is the assigning tasks part mentioned in the issue description but not included in this PR.

File: camel/tasks/task.py
Comment (willshang76) [inline camel/tasks/task.py:124]: It looks like that the relationship between parent task and sub tasks is not considered when we generate sub tasks.

File: camel/tasks/task.py
Comment (willshang76) [inline camel/tasks/task.py:179]: How about the dependency between the subtasks. I mean that there may be task 0.0 should be executed before task 0.1.

File: camel/tasks/task.py
Comment (willshang76) [inline camel/tasks/task.py:1]: One general question regarding tasks. For executing the tasks, when the subtasks are done, how could we gather all the results and feed back to the tasks depending on the sub tasks, in order for the agent/workforce to get a comprehensive solution for the top task.

File: examples/workforce/workforce.py
Comment (willshang76) [inline examples/workforce/workforce.py:45]: All the units share the same channel(queue), does it means all the messages in the queue are processed sequentially, even there may be tasks could be done parallel?

File: camel/utils/channel.py
Comment (willshang76) [inline camel/utils/channel.py:46]: In the example, all units share one channel and all try to get message out of the queue. If the workforce gets the message out of the queue which is not matched with the workforce id. Does it mean the queue message is gone without processed?

File: camel/workforce/workforce.py
Comment (yiyiyi0817) [reply camel/workforce/workforce.py:89]: My understanding is that `assign_other_workforce` is decided through the llm, but the details have not been worked out yet

File: camel/tasks/task.py
Comment (yiyiyi0817) [reply camel/tasks/task.py:1]: `Task` or `TaskManegr` module may need a `compose` function. I have talked about this with maoquan.

File: camel/tasks/task.py
Comment (yiyiyi0817) [reply camel/tasks/task.py:124]: For `id=f"{task_id}.{i}"`, `{task_id}` is the id of parent task and `.{i}` means the id of sub tasks.

File: camel/tasks/task.py
Comment (yiyiyi0817) [reply camel/tasks/task.py:179]: Maoquan said that it depend on the prompt of decompose task. Maybe we can treat all subtasks as if they were performed in sequence first.

File: camel/utils/channel.py
Comment (yiyiyi0817) [reply camel/utils/channel.py:46]: Yes, so from my dessign, every subtask should be assigned to a workforce as it is produced. Every workforce will only process tasks assign to itself.

File: examples/workforce/workforce.py
Comment (yiyiyi0817) [reply examples/workforce/workforce.py:45]: Tasks in the queue will be distributed to different units. They can process the tasks parallelly.

**Review by yiyiyi0817 — COMMENTED** [review]
Thanks for review.

**Comment by Appointat** [comment]
Hi, since this PR/issue is part of sprint 7, will you be opening the PR this week? (Just a reminder) : )

**Comment by yiyiyi0817** [comment]
> Hi, since this PR/issue is part of sprint 7, will you be opening the PR this week? (Just a reminder) : )

I think we will finish the development of this PR by the end of this week. It seems that it can be put on sprint7 or 8.

File: camel/workforce/workforce_prompt.py
Comment (Wendong-Fan) [inline camel/workforce/workforce_prompt.py:104]: this prompt could be further polished as I tested in examples/workforce/role_playing_with_agents.py the output is not so good

**Review by Wendong-Fan — COMMENTED** [review]
Thanks @yiyiyi0817  and @WHALEEYE !

The current implementation is task-solving oriented, I think in the long term we also need to support multi-agent communication for society simulation, like agents doing debate, we can explore this part later.

I feel it's not so necessary to set role_playing_workforce and single_agent_workforce just to convert roleplaying/agent into workforce, from user's side it's also complex to set up one workforce flow with current design, can we allow user input their agent, roleplaying, workforce objs directly into child_workforces list and do the transfer internally? WDYT?

File: camel/workforce/single_agent_workforce.py
Comment (Wendong-Fan) [inline camel/workforce/single_agent_workforce.py:27]: typo: SingleAgentWorkforce

File: camel/workforce/role_plying_workforce.py
Comment (Wendong-Fan) [inline camel/workforce/role_plying_workforce.py:34]: typo: RolePlayingWorkforce

File: camel/workforce/internal_workforce.py
Comment (yiyiyi0817) [reply camel/workforce/internal_workforce.py:86]: We designed this because we thought that it was rare for tasks to fail and that at least one main task needed to decompose initially to maximize the benefits of multiple agents.

File: camel/workforce/leaf_workforce.py
Comment (yiyiyi0817) [reply camel/workforce/leaf_workforce.py:32]: I think it can be added in next version. And we can make it easier for users to use parameters that combine chatagent and roleplaying. For example, let them configure all LeafWorkforce at once through yaml files or structured dictionaries

File: camel/workforce/workforce_prompt.py
Comment (yiyiyi0817) [reply camel/workforce/workforce_prompt.py:104]: Yes, I found that the summary formed by pormpt lost too much of the original information. Perhaps we need to feed prompt with examples of output, or make it output more and longer information.

**Review by yiyiyi0817 — COMMENTED** [review]
Thanks for your review! @Wendong-Fan

File: camel/workforce/internal_workforce.py
Comment (WHALEEYE) [reply camel/workforce/internal_workforce.py:121]: Should we merge this PR after structured output PR merged, or we can make a separate PR for this polish?

Comment (Wendong-Fan) [reply camel/workforce/internal_workforce.py]: strucuted output with openai models has been merged, you can refer to examples/structured_response/json_format_response.py, we can merge master and solve conflicts, then use this feature

**Comment by Appointat** [comment]
I helped resolve the conflicts in the pr withe master branch, including the issue of docstring format and pytest example.

**Review by Wendong-Fan — APPROVED** [review]
We can temporarily merge this PR and to the enhance further for the 2nd version as described here: https://github.com/camel-ai/camel/issues/796
The compose method of `Task` also need to be improved

**Pull request merged**
